runtime/compiler/optimizer/J9EstimateCodeSize.cpp

```diff
@@ -411,6 +415,132 @@ TR_J9EstimateCodeSize::adjustEstimateForMethodInvoke(TR_ResolvedMethod* method,
    return false;
    }
 
+bool
+TR_J9EstimateCodeSize::adjustEstimateForConstArgs(TR_CallTarget * target, int32_t& value, float factor)
+   {
+   static const char * disableConstArgWeightReduction = feGetEnv("TR_disableConstArgWeightReduction");
+   if (disableConstArgWeightReduction || !target->_calleeSymbol)
+      return false;
+
+   // This for remaining consistent with TR_MultipleCallTargetInliner::applyArgumentHeuristics
+   static const char * envKnownObjWeight = feGetEnv("TR_constClassWeight");
+   int32_t knownObjWeight = envKnownObjWeight ? atoi(envKnownObjWeight) : DEFAULT_KNOWN_OBJ_WEIGHT;
+
+   // The less aggressive adjustment factor, which is just the factor increased by 15% is
+   // intended to be used in cases where we can distinguish different degree of benefit
+   // to be had, with the original factor applied for the more beneficial cases, and the
+   // less aggressive factor being applied for the less beneficial cases.
+
+   float lessAggressiveAdjustmentFactor = factor * 1.15f;
+
+   int32_t originalWeight = value;
+   TR_LinkHead<TR_ParameterMapping> argMap;
+   TR_PrexArgInfo *prexArgInfo = target->_ecsPrexArgInfo;
+   if (((TR_J9InlinerPolicy *)_inliner->getPolicy())->validateArguments(target,argMap))
+      {
+      int32_t argIndex = 0;
+      for (TR_ParameterMapping* parm = argMap.getFirst(); parm; parm = parm->getNext())
+         {
+         int32_t interimWeight = value;
+         TR::Node * parmNode = parm->_parameterNode;
+         TR::ParameterSymbol * parmSym = parm->_parmSymbol;
+         const char * parmClassName = NULL;
+         int32_t parmClassNameLen = 0;
+         char * argClassName = NULL;
+         TR_PrexArgument * prexArg = NULL;
+         if (prexArgInfo) prexArg = prexArgInfo->get(argIndex++);
+
+         if (parmSym)
+            parmClassName = parmSym->getTypeSignature(parmClassNameLen);
+
+         if (prexArg && prexArg->getClass())
+            argClassName = TR::Compiler->cls.classSignature(comp(), prexArg->getClass(), comp()->trMemory());
+
+         if (parmNode->getSymbolReference()
+            && parmNode->getSymbolReference()->getSymbol()->isStatic()
+            && parmNode->getSymbolReference()->getSymbol()->castToStaticSymbol()->isConstString())
+            {
+            value *= factor;
+            heuristicTrace(tracer(),"Setting size from %d to %d because arg is constant string.", interimWeight, value);
+            }
+         else if (argClassName
+                 && parmClassName
+                 && strncmp(argClassName, parmClassName, parmClassNameLen) != 0 // arg type needs to be more specific than declared type to benefit from weight adjustment
+                 && strncmp(parmClassName, "Ljava/lang/String;", parmClassNameLen) == 0)
```

Comment:
This comparison (and the others like it) have been changed to compare against `parmClassName`, but I think `argClassName` was correct. If I'm not mistaken this will now match only when the parameter *is* declared to be of type `String` (and the argument is somehow not also known to be a `String`)

Changing this to compare with `argClassName` again, the `parmClassNameLen` isn't necessarily the right length, and we don't have the length of `argClassName` lying around. However, it looks like `argClassName` should be NUL-terminated, and the other comparand is a string literal, so the comparison could be done more conveniently using `strcmp` (no `n`)

Reply:
Fixed based on your suggestions.